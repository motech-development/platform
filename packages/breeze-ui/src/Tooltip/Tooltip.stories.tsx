import { withA11y } from '@storybook/addon-a11y';
import { select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import BaseStyles from '../BaseStyles/BaseStyles';
import Button from '../Button/Button';
import Tooltip from './Tooltip';

const stories = storiesOf('Tooltip', module);
const colour = {
  Danger: 'danger',
  Primary: 'primary',
};
const placement = {
  Bottom: 'bottom',
  Left: 'left',
  Right: 'right',
  Top: 'top',
};

stories.addDecorator(withA11y);
stories.addDecorator(withKnobs);

stories.add('Basic tooltip', () => (
  <>
    <BaseStyles />

    <Tooltip
      id="test"
      parent={
        <Button colour="primary" size="lg">
          Hover over me
        </Button>
      }
      placement={
        select('Placement', placement, 'bottom') as
          | 'bottom'
          | 'left'
          | 'right'
          | 'top'
      }
      colour={select('Colour', colour, 'primary')}
      message={text('Text', 'This is a tooltip!')}
    />
  </>
));
