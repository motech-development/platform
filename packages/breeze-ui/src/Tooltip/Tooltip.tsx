import React, { FC, memo, ReactNode, useEffect, useRef, useState } from 'react';
import { Manager, Popper, Reference } from 'react-popper';
import styled, { ThemeProvider } from 'styled-components';

const tooltipTheme = {
  danger: {
    background: 'rgb(199,56,79)',
    colour: '#fff',
  },
  primary: {
    background: '#007fa8',
    colour: '#fff',
  },
  secondary: {
    background: '#f6f9fc',
    colour: '#333',
  },
  success: {
    background: 'rgb(0,128,93)',
    colour: '#fff',
  },
};

type TooltipPlacement =
  | 'auto-start'
  | 'auto'
  | 'auto-end'
  | 'top-start'
  | 'top'
  | 'top-end'
  | 'right-start'
  | 'right'
  | 'right-end'
  | 'bottom-end'
  | 'bottom'
  | 'bottom-start'
  | 'left-end'
  | 'left'
  | 'left-start';

interface ITooltipPlacement {
  $colour: keyof typeof tooltipTheme;
  $placement: TooltipPlacement;
}

const TooltipArrow = styled.div<ITooltipPlacement>`
  ${({ $colour, $placement, theme }) => `
    height: 0;
    position: absolute;
    width: 0;

    ${(() => {
      switch ($placement) {
        case 'bottom':
          return `
            border-bottom: 5px solid ${theme[$colour].background};
            border-left: 5px solid transparent;
            border-right: 5px solid transparent;
            top: -5px;
          `;
        case 'left':
          return `
            border-bottom: 5px solid transparent;
            border-left: 5px solid ${theme[$colour].background};
            border-top: 5px solid transparent;
            right: -5px;
          `;
        case 'right':
          return `
            border-bottom: 5px solid transparent;
            border-right: 5px solid ${theme[$colour].background};
            border-top: 5px solid transparent;
            left: -5px;
          `;
        case 'top':
          return `
            border-left: 5px solid transparent;
            border-right: 5px solid transparent;
            border-top: 5px solid ${theme[$colour].background};
            bottom: -5px;
          `;
        default:
          return '';
      }
    })()}
  `}
`;

const TooltipWrapper = styled.div`
  display: inline-block;
  position: relative;
`;

interface ITooltipContent extends ITooltipPlacement {
  $visible: boolean;
}

const TooltipContent = styled.div<ITooltipContent>`
  ${({ $colour, $visible, theme }) => `
    background-color: ${theme[$colour].background};
    color: ${theme[$colour].colour};
    font-size: 14px;
    line-height: 22px;
    padding: 0 5px;
    visibility: ${$visible ? 'visible' : 'hidden'};
  `}
`;

export interface ITooltipProps {
  colour?: keyof typeof tooltipTheme;
  id: string;
  message: ReactNode;
  parent: ReactNode;
  placement: TooltipPlacement;
}

const Tooltip: FC<ITooltipProps> = ({
  colour = 'primary',
  message,
  id,
  parent,
  placement,
}) => {
  let timer: number;
  const [visible, setVisibility] = useState(false);
  const mounted = useRef(false);
  const hideTooltip = () => {
    timer = setTimeout(() => {
      if (mounted.current) {
        setVisibility(false);
      }
    }, 1000);
  };
  const showTooltip = () => {
    clearTimeout(timer);
    setVisibility(true);
  };

  useEffect(() => {
    mounted.current = true;

    return () => {
      mounted.current = false;
    };
  });

  return (
    <ThemeProvider theme={tooltipTheme}>
      <Manager>
        <Reference>
          {({ ref }) => (
            <TooltipWrapper
              ref={ref}
              onBlur={hideTooltip}
              onFocus={showTooltip}
              onMouseOver={showTooltip}
              onMouseOut={hideTooltip}
            >
              {parent}
            </TooltipWrapper>
          )}
        </Reference>

        <Popper
          placement={placement}
          strategy="fixed"
          modifiers={[
            {
              name: 'offset',
              options: {
                offset: [0, 5],
              },
            },
          ]}
        >
          {({ arrowProps, placement: popperPlacement, ref, style }) => {
            const placementToUse =
              process.env.NODE_ENV === 'test' ? placement : popperPlacement;

            return (
              <TooltipContent
                $placement={placementToUse}
                $visible={visible}
                id={id}
                role="tooltip"
                ref={ref}
                $colour={colour}
                style={style}
              >
                {message}
                <TooltipArrow
                  $placement={placementToUse}
                  ref={arrowProps.ref}
                  $colour={colour}
                  style={arrowProps.style}
                />
              </TooltipContent>
            );
          }}
        </Popper>
      </Manager>
    </ThemeProvider>
  );
};

export default memo(Tooltip);
