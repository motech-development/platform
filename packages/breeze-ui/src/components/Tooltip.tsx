import { Placement } from '@popperjs/core';
import { FC, ReactNode, useEffect, useRef, useState } from 'react';
import { usePopper } from 'react-popper';
import { classNames, themeClass } from '../utils/className';
import TTheme from '../utils/theme';

export interface ITooltipProps {
  colour?: TTheme;
  id: string;
  message: ReactNode;
  parent: ReactNode;
  placement?: Placement;
}

const Tooltip: FC<ITooltipProps> = ({
  colour = 'primary',
  message,
  id,
  parent,
  placement = 'auto',
}) => {
  let timer: ReturnType<typeof setTimeout>;
  const mounted = useRef(false);
  const [visible, setVisibility] = useState(false);
  const [referenceElement, setReferenceElement] =
    useState<HTMLDivElement | null>();
  const [popperElement, setPopperElement] = useState<HTMLDivElement | null>();
  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 10],
        },
      },
    ],
    placement,
  });
  const hideTooltip = () => {
    timer = setTimeout(() => {
      if (mounted.current) {
        setVisibility(false);
      }
    }, 1000);
  };
  const showTooltip = () => {
    clearTimeout(timer);

    setVisibility(true);
  };

  useEffect(() => {
    mounted.current = true;

    return () => {
      mounted.current = false;
    };
  }, []);

  return (
    <>
      <div
        className="inline-block"
        onBlur={hideTooltip}
        onFocus={showTooltip}
        onMouseOver={showTooltip}
        onMouseOut={hideTooltip}
        ref={setReferenceElement}
      >
        {parent}
      </div>

      {/* @tailwind: bg-blue-600 bg-gray-100 bg-green-600 bg-red-600 bg-yellow-600 */}
      {visible && (
        <div
          id={id}
          className={classNames(
            'p-1 text-sm font-display rounded-sm',
            themeClass(colour, 'bg-{theme}-600 text-white', {
              secondary: 'bg-{theme}-100 text-gray-600',
            }),
          )}
          ref={setPopperElement}
          role="tooltip"
          style={styles.popper}
          {...attributes.popper}
        >
          <p className="whitespace-nowrap">{message}</p>
        </div>
      )}
    </>
  );
};

export default Tooltip;
