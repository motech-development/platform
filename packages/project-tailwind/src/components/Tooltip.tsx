import {
  arrow,
  autoUpdate,
  flip,
  MiddlewareData,
  offset,
  Placement,
  shift,
  Side,
  useDismiss,
  useFloating,
  useFocus,
  useHover,
  useInteractions,
  useRole,
} from '@floating-ui/react';
import {
  ComponentPropsWithoutRef,
  ReactNode,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Themes, TTheme, useTailwind } from '../utilities/tailwind';

/** Default legth of time to display tooltip after moving away from parent */
const DEFAULT_TIME = 1000;

/** Arrow box dimension */
const BOX_SIZE = 8;

/** Arrow box's hypotenuse length */
const FLOATING_OFFSET = Math.sqrt(2 * BOX_SIZE ** 2) / 2;

/**
 * Creates inline styles for arrows
 *
 * @param element - Arrow from middleware
 *
 * @param placement - Arrow placement
 *
 * @returns Arrow inline styles
 */
function createArrowStyles(
  element: MiddlewareData['arrow'],
  placement: Placement,
) {
  let styles = {
    bottom: '',
    left: '',
    right: '',
    top: '',
  };

  if (element) {
    const { x, y } = element;

    const side = placement.split('-')[0];

    /** Position to apply generated style */
    const xy = {
      bottom: 'top',
      left: 'right',
      right: 'left',
      top: 'bottom',
    }[side] as string;

    styles = {
      ...styles,
      left: xy === 'right' ? '' : `${x ?? BOX_SIZE}px`,
      top: xy === 'bottom' ? '' : `${y ?? BOX_SIZE}px`,
      [xy]: -BOX_SIZE / 2,
    };
  }

  return styles;
}

/** Tooltip component properties */
export interface ITooltipProps extends ComponentPropsWithoutRef<'div'> {
  /** Content to be displayed */
  content: ReactNode;

  /** Callback trigger when tooltop is shown or hidden  */
  onVisibilityChange?: (value: boolean) => void;

  /** Element to output tooltip relative to */
  parent: ReactNode;

  /** Where tooltip should be placed */
  position?: Side;

  /** Length of time to keep tooltip visible in milliseconds */
  time?: number;

  /** Component theme */
  theme?: TTheme;
}

/**
 * Display information when hovering over a component
 *
 * @param props - Component props
 *
 * @returns Tooltip component
 */
export function Tooltip({
  className,
  content,
  onVisibilityChange,
  parent,
  position = 'bottom',
  time = DEFAULT_TIME,
  theme = Themes.PRIMARY,
  ...rest
}: ITooltipProps) {
  const arrowRef = useRef(null);

  const { createStyles } = useTailwind(theme);

  const [isOpen, setIsOpen] = useState(false);

  const { context, middlewareData, placement, refs, strategy, x, y } =
    useFloating({
      middleware: [
        flip(),
        offset(FLOATING_OFFSET),
        shift(),
        arrow({
          element: arrowRef,
        }),
      ],
      onOpenChange: setIsOpen,
      open: isOpen,
      placement: position,
      whileElementsMounted: autoUpdate,
    });

  const hover = useHover(context, {
    delay: {
      close: time,
    },
    move: false,
  });

  const focus = useFocus(context);

  const dismiss = useDismiss(context);

  const role = useRole(context, {
    role: 'tooltip',
  });

  const { getFloatingProps, getReferenceProps } = useInteractions([
    hover,
    focus,
    dismiss,
    role,
  ]);

  const tooltipStyles = createStyles({
    classNames: [className],
    theme: {
      danger: ['bg-red-600 text-red-50'],
      primary: ['bg-blue-600 text-blue-50'],
      secondary: ['bg-gray-100 text-gray-600'],
      success: ['bg-green-600 text-green-50'],
      warning: ['bg-yellow-600 text-yellow-50'],
    },
  });

  const innerTooltipStyles = createStyles({
    classNames: [
      'px-1 text-sm font-display whitespace-nowrap',
      {
        'border-b-2': placement === 'top',
        'border-l-2': placement === 'right',
        'border-r-2': placement === 'left',
        'border-t-2': placement === 'bottom',
      },
    ],
    theme: {
      danger: ['border-red-700'],
      primary: ['border-blue-700'],
      secondary: ['border-gray-200'],
      success: ['border-green-700'],
      warning: ['border-yellow-700'],
    },
  });

  const tooltipInlineStyles = {
    left: x ?? 0,
    position: strategy,
    top: y ?? 0,
    width: 'max-content',
  };

  const arrowStyles = createStyles({
    classNames: ['absolute w-2 h-2 rotate-45 -z-10 pointer-events-none'],
    theme: {
      danger: ['bg-red-700'],
      primary: ['bg-blue-700'],
      secondary: ['bg-gray-200'],
      success: ['bg-green-700'],
      warning: ['bg-yellow-700'],
    },
  });

  const arrowInlineStyles = createArrowStyles(middlewareData.arrow, placement);

  useEffect(() => {
    if (onVisibilityChange) {
      onVisibilityChange(isOpen);
    }
  }, [isOpen, onVisibilityChange]);

  return (
    <>
      <div
        className="inline-block"
        data-testid="tooltip-parent-element"
        ref={refs.setReference}
        {...getReferenceProps()}
      >
        {parent}
      </div>

      {isOpen && (
        <div
          className={tooltipStyles}
          ref={refs.setFloating}
          style={tooltipInlineStyles}
          {...getFloatingProps()}
          {...rest}
        >
          <div className={innerTooltipStyles}>{content}</div>

          <div
            className={arrowStyles}
            ref={arrowRef}
            style={arrowInlineStyles}
          />
        </div>
      )}
    </>
  );
}
