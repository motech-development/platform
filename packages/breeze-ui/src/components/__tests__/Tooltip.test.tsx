import { act, fireEvent, render } from '@testing-library/react';
import Tooltip from '../Tooltip';

describe('Tooltip', () => {
  let button: HTMLElement;
  let tooltip: HTMLElement;

  describe('visibility', () => {
    it('should hide after 1000ms', async () => {
      const { findByTestId } = render(
        <Tooltip
          id="test"
          parent={
            <button type="button" data-testid="button">
              Hello world
            </button>
          }
          placement="top"
          message="Hello"
        />,
      );

      button = await findByTestId('button');

      jest.useFakeTimers();

      fireEvent.blur(button);

      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000);
    });

    it('should hide the tooltip', async () => {
      jest.useFakeTimers();

      const { findByTestId, queryByRole } = render(
        <Tooltip
          id="test"
          parent={
            <button type="button" data-testid="button">
              Hello world
            </button>
          }
          placement="top"
          message="Hello"
        />,
      );

      button = await findByTestId('button');

      fireEvent.focus(button);

      expect(queryByRole('tooltip')).toBeDefined();

      fireEvent.blur(button);

      act(() => {
        jest.runOnlyPendingTimers();
      });

      expect(queryByRole('tooltip')).toBeNull();
    });
  });

  describe('primary', () => {
    beforeEach(async () => {
      const { findByRole, findByTestId } = render(
        <Tooltip
          id="test"
          parent={
            <button type="button" data-testid="button">
              Hello world
            </button>
          }
          placement="top"
          message="Hello"
        />,
      );

      button = await findByTestId('button');

      fireEvent.focus(button);

      tooltip = await findByRole('tooltip');
    });

    it('should show the tooltip on hover', () => {
      expect(tooltip).toBeDefined();
    });

    it('should have the correct styles', () => {
      expect(tooltip).toMatchSnapshot();
    });

    it('should display content', () => {
      const { textContent } = tooltip.firstChild as ChildNode;

      expect(textContent).toEqual('Hello');
    });
  });

  describe('danger', () => {
    beforeEach(async () => {
      const { findByRole, findByTestId } = render(
        <Tooltip
          id="test"
          parent={
            <button type="button" data-testid="button">
              Hello world
            </button>
          }
          placement="top"
          message="Hello"
          colour="danger"
        />,
      );

      button = await findByTestId('button');

      fireEvent.focus(button);

      tooltip = await findByRole('tooltip');
    });

    it('should show the tooltip on hover', () => {
      expect(tooltip).toBeDefined();
    });

    it('should have the correct styles', () => {
      expect(tooltip).toMatchSnapshot();
    });
  });

  describe('secondary', () => {
    beforeEach(async () => {
      const { findByRole, findByTestId } = render(
        <Tooltip
          id="test"
          parent={
            <button type="button" data-testid="button">
              Hello world
            </button>
          }
          placement="top"
          message="Hello"
          colour="secondary"
        />,
      );

      button = await findByTestId('button');

      fireEvent.focus(button);

      tooltip = await findByRole('tooltip');
    });

    it('should show the tooltip on hover', () => {
      expect(tooltip).toBeDefined();
    });

    it('should have the correct styles', () => {
      expect(tooltip).toMatchSnapshot();
    });
  });

  describe('success', () => {
    beforeEach(async () => {
      const { findByRole, findByTestId } = render(
        <Tooltip
          id="test"
          parent={
            <button type="button" data-testid="button">
              Hello world
            </button>
          }
          placement="top"
          message="Hello"
          colour="success"
        />,
      );

      button = await findByTestId('button');

      fireEvent.focus(button);

      tooltip = await findByRole('tooltip');
    });

    it('should show the tooltip on hover', () => {
      expect(tooltip).toBeDefined();
    });

    it('should have the correct styles', () => {
      expect(tooltip).toMatchSnapshot();
    });
  });

  describe('warning', () => {
    beforeEach(async () => {
      const { findByRole, findByTestId } = render(
        <Tooltip
          id="test"
          parent={
            <button type="button" data-testid="button">
              Hello world
            </button>
          }
          placement="top"
          message="Hello"
          colour="warning"
        />,
      );

      button = await findByTestId('button');

      fireEvent.focus(button);

      tooltip = await findByRole('tooltip');
    });

    it('should show the tooltip on hover', () => {
      expect(tooltip).toBeDefined();
    });

    it('should have the correct styles', () => {
      expect(tooltip).toMatchSnapshot();
    });
  });
});
