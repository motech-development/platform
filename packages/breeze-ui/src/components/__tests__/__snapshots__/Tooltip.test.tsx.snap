// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip danger should have the correct styles 1`] = `
<div
  class="p-1 text-sm font-display rounded-sm bg-red-600 text-white"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  id="test"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -10px);"
>
  <p
    class="whitespace-nowrap"
  >
    Hello
  </p>
</div>
`;

exports[`Tooltip primary should have the correct styles 1`] = `
<div
  class="p-1 text-sm font-display rounded-sm bg-blue-600 text-white"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  id="test"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -10px);"
>
  <p
    class="whitespace-nowrap"
  >
    Hello
  </p>
</div>
`;

exports[`Tooltip secondary should have the correct styles 1`] = `
<div
  class="p-1 text-sm font-display rounded-sm bg-gray-100 text-gray-600"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  id="test"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -10px);"
>
  <p
    class="whitespace-nowrap"
  >
    Hello
  </p>
</div>
`;

exports[`Tooltip success should have the correct styles 1`] = `
<div
  class="p-1 text-sm font-display rounded-sm bg-green-600 text-white"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  id="test"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -10px);"
>
  <p
    class="whitespace-nowrap"
  >
    Hello
  </p>
</div>
`;

exports[`Tooltip warning should have the correct styles 1`] = `
<div
  class="p-1 text-sm font-display rounded-sm bg-yellow-600 text-white"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  id="test"
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -10px);"
>
  <p
    class="whitespace-nowrap"
  >
    Hello
  </p>
</div>
`;
